python/ray/serve/pipeline/generate.py

```diff
@@ -129,3 +132,29 @@ def extractor(dag_node):
     serve_dag_root._apply_recursive(extractor)
 
     return list(deployments.values())
+
+
+def get_ingress_deployment(
+    serve_dag_root_node: DAGNode, pipeline_input_node: PipelineInputNode
+) -> Deployment:
+    """Return an Ingress deployment to handle user HTTP inputs.
+
+    Args:
+        serve_dag_root_node (DAGNode): Transformed  as serve DAG's root. User
+            inputs are translated to serve_dag_root_node.execute().
+        pipeline_input_node (DAGNode): Singleton PipelineInputNode instance that
+            contains input preprocessor info.
+    Returns:
+        ingress (Deployment): Generated pipeline ingress deployment to serve
+            user HTTP requests.
+    """
+    serve_dag_root_json = json.dumps(serve_dag_root_node, cls=DAGNodeEncoder)
+    preprocessor_import_path = pipeline_input_node.get_preprocessor_import_path()
+    serve_dag_root_deployment = Ingress.options(
+        init_args=(
+            serve_dag_root_json,
+            preprocessor_import_path,
+        )
+    )
```

Comment:
Hmm I just realized that this is going to be somewhat problematic because serve names are case-sensitive... so people will need to access this like `app.Ingress`. This might be ok, but maybe better to make them case-insensitive

Reply:
given the `Ingress` deployment is not directly written or visible to user .. how about we just do a simple conversion for `app.ingress` that ensures 1) it's not taken by user in DAG binding 2) it maps to the corresponding `Ingress` deployment ? I think it's cleaner if this attr access is done at app level.

Reply:
I don't fully understand your suggestion, but we shouldn't special-case anything in the `app`. We can address my concern by just setting `name="ingress"` in this deployment

Reply:
that could work for ingress case. I think we need to resolve a mapping issue for serve application given it can work with deployment objects directly since deployment names are unique;

In serve pipeline `build` we would carry new info from user such as variable name, for a DAG like 
```

with ServeInputNode(preprocessor=request_to_data_int) as dag_input:
    m1 = Model.bind(1)
    m2 = Model.bind(2)
    m1_output = m1.forward.bind(dag_input[0])
    m2_output = m2.forward.bind(dag_input[1])
    ray_dag = ensemble.bind(m1_output, m2_output, dag_input[2])
```
users are using `m1` and `m2` in DAG for multiple instantiation of the same `Model` class, which we generate `Model` and `Model_1` as deployment names.

So we'll end up something like 
```
{
  "m1": Deployment(name="Model")
  "m2": Deployment(name="Model_1")
  "ingress": Deployment(name="Ingress")
}
```

that `app.m1` needs to map to `Deployment(name="Model")`. Alternatively we can also
1) Enforce deployment name to always be same as variable name, we end up with "m1": Deployment(name="m1"), it's not as consistent as our current behavior that we use class name as default deployment name. (It's slightly less of a problem modifying YAML since it has import path)

2) Pass a mapping to `Serve Application` so it knows how to map to the Deployment object with upper case class name.

3) Do `app.Model_1` .. that sounds to be least appealing to me.

Reply:
Either way there's no harm setting it "ingress" name here and we and discuss the options above on slack channel.